File: apps/desktop/src/litTheme.ts
import { activityDotKey, type DotKey } from './activity';
import { isRecentlyActive } from './desktopNavigation';
import type { UiTheme } from './preferences';
import type { DrawerSession } from './types';

/**
 * `none` is not "idle": a stopped session, or no session at all, must not tint
 * the app warm or cool — a dead shell that keeps glowing reads as a live one.
 */
export type LitState = 'working' | 'waiting' | 'done' | 'idle' | 'none';

/** Classified exactly as the drawer classifies its own rows, so the row and
 * the chrome can never disagree about what a session is doing. */
export function activeSessionDot(
  sessions: DrawerSession[],
  hostId: string | null,
  sessionId: string,
): { session: DrawerSession | null; dot: DotKey | null } {
  const session = sessions.find((row) => row.hostId === hostId && row.id === sessionId) ?? null;
  if (!session) return { session: null, dot: null };
  return {
    session,
    dot: activityDotKey(session.status, session.activity, isRecentlyActive(session.last_output_at)),
  };
}

/** Reuses the drawer's own classification so the row and the chrome can never disagree. */
export function litStateFor(dot: DotKey | null): LitState {
  switch (dot) {
    case 'working':
      return 'working';
    case 'waiting':
      return 'waiting';
    case 'done':
      return 'done';
    case 'idle':
      return 'idle';
    default:
      // 'stopped' and null both mean "nothing is running here".
      return 'none';
  }
}

/** Per-state rather than one shared alpha: ember reads much hotter than amber
 * at parity, which made waiting look like an alarm rather than a state. */
const BLOOM: Record<LitState, { b1: string; b2: string; b3: string; rim: string }> = {
  working: { b1: '13%', b2: '6%', b3: '2%', rim: '55%' },
  waiting: { b1: '9%', b2: '4%', b3: '1.5%', rim: '46%' },
  // Between waiting and idle: warmer than a shell that is merely alive, quieter
  // than one that is blocked. Finishing is worth noticing and nothing more.
  done: { b1: '8%', b2: '3.5%', b3: '1.2%', rim: '36%' },
  idle: { b1: '7%', b2: '3%', b3: '1%', rim: '30%' },
  none: { b1: '0%', b2: '0%', b3: '0%', rim: '0%' },
};

export function litColor(theme: UiTheme, state: LitState): string {
  switch (state) {
    case 'working':
      return theme.heat.working;
    case 'waiting':
      return theme.heat.waiting;
    case 'done':
      return theme.heat.done;
    case 'idle':
      return theme.heat.cool;
    case 'none':
      return theme.colors.textFaint;
  }
}

/** Matches the `heat-arrive` keyframe in `index.css` plus slack so the
 * attribute outlives the animation rather than cutting it off. */
export const ARRIVAL_MS = 800;

/** Fires only on ENTERING waiting, since the drawer re-reports state every poll
 * and `settled` keeps the swell from firing on startup's initial `none → waiting`. */
export function shouldAnnounceArrival(prev: LitState, next: LitState, settled: boolean): boolean {
  return settled && next === 'waiting' && prev !== 'waiting';
}

/** Everything tinted in `index.css` resolves through these, so re-tinting the
 * app on a session switch is four variable writes. */
export function litVars(theme: UiTheme, state: LitState): Record<string, string> {
  const bloom = BLOOM[state];
  return {
    '--lit': litColor(theme, state),
    '--b1': bloom.b1,
    '--b2': bloom.b2,
    '--b3': bloom.b3,
    '--rim': bloom.rim,
  };
}

/** Every token the flavour defines is published — the previous version emitted
 * nine of sixteen, so many `index.css` rules carried a hard-coded fallback. */
export function shellVars(theme: UiTheme, state: LitState): Record<string, string> {
  const c = theme.colors;
  return {
    '--background': c.background,
    '--surface': c.surface,
    '--surface-raised': c.surfaceRaised,
    '--input': c.input,
    '--text': c.text,
    '--text-muted': c.textMuted,
    '--text-faint': c.textFaint,
    '--border': c.border,
    '--overlay': c.overlay,
    '--selected': c.selected,
    '--accent': c.accent,
    '--accent-text': c.accentText,
    '--success': c.success,
    '--warning': c.warning,
    '--danger': c.danger,
    '--info': c.info,
    // So padding around the emulator matches it; `--input` was close enough to
    // be invisible in dark but showed up as a pale gutter in light flavours.
    '--term-bg': theme.terminal.background,
    ...litVars(theme, state),
  };
}

File: clients/apple/TetherKit/Sources/TetherKit/Theme/LitTheme.swift
import Foundation
import SwiftUI

/// What the chrome is wearing. `none` is not "idle": a stopped or absent session
/// must not tint the app, or a dead shell reads as a live one.
public enum LitState: String, Equatable, Sendable {
  case working
  case waiting
  case done
  case idle
  case none
}

/// Bloom alphas for the glow, per-state: ember reads hotter than amber at equal
/// alpha, so waiting is scaled quieter; idle quieter still; none zeroed.
public struct LitBloom: Equatable, Sendable {
  public var b1: Double
  public var b2: Double
  public var b3: Double
  public var rim: Double

  public static let working = LitBloom(b1: 0.13, b2: 0.06, b3: 0.02, rim: 0.55)
  public static let waiting = LitBloom(b1: 0.09, b2: 0.04, b3: 0.015, rim: 0.46)
  // Between waiting and idle: warmer than a shell that is merely alive, quieter
  // than one that is blocked. Finishing is worth noticing and nothing more.
  public static let done = LitBloom(b1: 0.08, b2: 0.035, b3: 0.012, rim: 0.36)
  public static let idle = LitBloom(b1: 0.07, b2: 0.03, b3: 0.01, rim: 0.30)
  public static let none = LitBloom(b1: 0, b2: 0, b3: 0, rim: 0)

  public static func forState(_ state: LitState) -> LitBloom {
    switch state {
    case .working: .working
    case .waiting: .waiting
    case .done: .done
    case .idle: .idle
    case .none: .none
    }
  }
}

/// Pure classification + colour resolution for Aurora's lit chrome.
public enum LitTheme {
  /// Reuses the drawer's own classification so the row and the chrome can never disagree.
  public static func state(for dot: SessionActivityDot?) -> LitState {
    switch dot {
    case .working: .working
    case .waiting: .waiting
    case .done: .done
    case .idle: .idle
    case .stopped, .none: .none
    }
  }

  public static func state(
    status: String?,
    activity: String?,
    lastOutputAt: String?
  ) -> LitState {
    guard let status else { return .none }
    let live = SessionActivityLogic.isRecentlyActive(lastOutputAt: lastOutputAt)
    let dot = SessionActivityLogic.dotKey(status: status, activity: activity, live: live)
    return state(for: dot)
  }

  public static func color(for state: LitState) -> Color {
    switch state {
    case .working: TetherColors.heatWorking
    case .waiting: TetherColors.heatWaiting
    case .done: TetherColors.heatDone
    case .idle: TetherColors.heatCool
    case .none: TetherColors.textFaint
    }
  }

  public static func label(for state: LitState) -> String {
    switch state {
    case .working: "working"
    case .waiting: "waiting"
    case .done: "finished"
    case .idle: "idle"
    case .none: "stopped"
    }
  }
}

/// Runtime chrome tint published through the SwiftUI environment.
public struct LitChrome: Equatable {
  public var state: LitState
  public var color: Color
  public var bloom: LitBloom

  public static let none = LitChrome(
    state: .none,
    color: TetherColors.textFaint,
    bloom: .none
  )

  public static func resolve(
    status: String?,
    activity: String?,
    lastOutputAt: String?
  ) -> LitChrome {
    let state = LitTheme.state(
      status: status,
      activity: activity,
      lastOutputAt: lastOutputAt
    )
    return LitChrome(
      state: state,
      color: LitTheme.color(for: state),
      bloom: LitBloom.forState(state)
    )
  }
}

private struct LitChromeKey: EnvironmentKey {
  static let defaultValue = LitChrome.none
}

extension EnvironmentValues {
  public var litChrome: LitChrome {
    get { self[LitChromeKey.self] }
    set { self[LitChromeKey.self] = newValue }
  }
}
